src/states/paused.rs

```diff
@@ -1,22 +1,53 @@
-use amethyst::{input::is_key_down, prelude::*, renderer::VirtualKeyCode};
+use amethyst::{
+    input::InputEvent,
+    prelude::*,
+    shrev::{EventChannel, ReaderId},
+};
 
-pub struct PausedState;
+pub struct PausedState {
+    input_event_reader_id: Option<ReaderId<InputEvent<String>>>,
+}
+
+impl Default for PausedState {
+    fn default() -> Self {
+        PausedState {
+            input_event_reader_id: None,
+        }
+    }
+}
 
 impl SimpleState for PausedState {
-    fn handle_event(
-        &mut self,
-        _data: StateData<'_, GameData<'_, '_>>,
-        event: StateEvent,
-    ) -> SimpleTrans {
-        if let StateEvent::Window(event) = &event {
-            if is_key_down(&event, VirtualKeyCode::Escape) {
-                return Trans::Pop;
-            }
-            if is_key_down(&event, VirtualKeyCode::P) {
-                return Trans::Pop;
-            }
+    fn on_start(&mut self, data: StateData<'_, GameData<'_, '_>>) {
+        {
+            let mut input_event_channel = data
+                .world
+                .write_resource::<EventChannel<InputEvent<String>>>();
+            self.input_event_reader_id = Some(input_event_channel.register_reader());
         }
+    }
+
+    fn on_resume(&mut self, data: StateData<'_, GameData<'_, '_>>) {
+        {
```

Comment:
There are a couple of unnecessary scopes, could you remove them?

Reply:
Should be OK now.

Reply:
I meant the additional braces in line `21` and `26` - but it's good to know that it is also possible to remove the lifetime definitions :)

Reply:
Oh, the additional braces are here on purpose. If we want to use the `data.world` for something else later on, we need them to end the mutable borrow.

Reply:
Although I can just rewrite it so that the mutable borrow ends directly.

Reply:
Good point, always code with change in mind 🙏